File: day17/task1.erl
-module(task1).
-export([main/0]).

-spec powerset(list(T)) -> list(list(T)).
powerset([Head | Tail]) ->
  Without = powerset(Tail),
  With = lists:map(fun (Elem) -> [Head | Elem] end, Without),
  Without ++ With;
powerset([]) -> [[]].

-spec solution(integer(), list(integer())) -> integer().
solution(Liters, Containers) ->
  S = powerset(Containers),
  Sums = lists:map(fun lists:sum/1, S),
  Proper = lists:filter(fun (X) -> X =:= Liters end, Sums),
  length(Proper).

-spec read_containers() -> list(integer()).
read_containers() ->
  {ok, Binary} = file:read_file("input.txt"),
  Parts = binary:split(string:trim(Binary), <<"\n">>, [global]),
  lists:map(fun binary_to_integer/1, Parts).

-spec main() -> none().
main() ->
  {ok, Liters} = io:read("Liters: "),
  Containers = read_containers(),
  Answer = solution(Liters, Containers),
  io:format("Answer is: ~p\n", [Answer]).

File: day17/task1.lua
local function map(table, transform)
  local output = {}
  for index, value in pairs(table) do
    output[index] = transform(value, index)
  end
  return output
end

local function filter(table, selector)
  local output = {}
  for index, value in ipairs(table) do
    if selector(value, index) then
      output[#output + 1] = value
    end
  end
  return output
end

local function sum(table)
  local output = 0
  for _, each in pairs(table) do
    output = output + each
  end
  return output
end

local function concat_arrays(a, b)
  local output = {}
  for _, each in ipairs(a) do
    output[#output + 1] = each
  end
  for _, each in ipairs(b) do
    output[#output + 1] = each
  end
  return output
end

local function powerset(source)
  if #source <= 0 then
    return {{}}
  end
  local head = source[1]
  local tail = { select(2, table.unpack(source)) }
  local without = powerset(tail)
  local with = map(without, function (elem)
    return {head, table.unpack(elem)}
  end)
  return concat_arrays(without, with)
end

local function solution(liters, containers)
  local S = powerset(containers)
  local sums = map(S, sum)
  return #filter(sums, function (elem)
    return elem == liters
  end)
end

local function read_containers(filename)
  local containers = {}
  for line in io.lines(filename) do
    containers[#containers + 1] = tonumber(line, 10)
  end
  return containers
end


local function main ()
  io.write("Liters: ")
  local liters = io.read("n*")
  local containers = read_containers("input.txt")
  local answer = solution(liters, containers)
  io.write("Answer: " .. answer .. "\n")
end

main()
